### PR TITLE
feat: More dtypes supports cast to list

### LINKED ISSUE #10709: Add support for cast `LargeUtf8` to `LargeList(LargeUtf8)`
### Problem description

I have a huge dataset (almost million JSONs) containing a lot of nested fields that can either have String Dtype or List(String) Dtype. Ideally everything should be in this dataset a list of Strings. Currently there is no quick way of casting all these fields automatically to List(string). So I need to load everything with a lot of custom Struct schema's and then using `.str.json_extract` :

Small example of the current workflow:
```python
df = pl.DataFrame(
    pl.Series("dict", [json.dumps(item) for item in data["dicts"]])
)

df.with_columns(
    pl.col("dict").str.json_extract(struct_schema_default).alias("parsed_dict_utf"), # This reads everything as UTF8
    pl.col("dict").str.json_extract(struct_schema_list_version).alias("parsed_dict_list"), # This reads  everything as List(Utf8)
)
```
Then there are a lot of coalesce and concat_list steps to combine the UTF8 and List(Utf8) columns.

But if the below casting would be supported,
```python
ArrowErrorException: NotYetImplemented("Casting from LargeUtf8 to LargeList(Field { name: \"item\", data_type: LargeUtf8, is_nullable: true, metadata: {} }) not supported")
```
Then it could be simplified very much because I only need one struct schema instead of 15+ schema's. To parse all fields that contain Utf8 automatically to LargeList(Utf8)

### PR DESCRIPTION
**Pull request by reswqa**
This closes #10709.

Upstream pull request: https://github.com/jorgecarleitao/arrow2/pull/1567. Please review it first. :)

I will change the `arrow2` dependency to the correct upstream `rev` after that being merged.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -54,8 +54,8 @@ xxhash-rust = { version = "0.8.6", features = ["xxh3"] }
 
 [workspace.dependencies.arrow]
 package = "arrow2"
-git = "https://github.com/jorgecarleitao/arrow2"
-rev = "ba6a882bc1542b0b899774b696ebea77482b5c31"
+git = "https://github.com/reswqa/arrow2"
+branch = "to_large_list"
```
Comment (reswqa) [inline Cargo.toml:58]: I will change the arrow2 dependency to the correct upstream rev after that being merged.

Comment (ritchie46) [reply Cargo.toml:58]: Merged upstream. Can you rebease?

Comment (reswqa) [reply Cargo.toml]: Sure, rebased and updated to new rev.

**Comment by ritchie46** [comment]
Alright! Thanks.

**Comment by ion-elgreco** [comment]
@reswqa Thanks a lot!

**Pull request merged**